sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/datasketchesAggregates.scala

```diff
@@ -0,0 +1,336 @@
+    inputAggBufferOffset: Int = 0)
+  extends TypedImperativeAggregate[Union] with BinaryLike[Expression] with ExpectsInputTypes {
+
+  // Union config - mark as lazy so that they're not evaluated during tree transformation.
+
+  lazy val lgMaxK: Int = lgMaxKExpression.eval().asInstanceOf[Int]
+
+  // Constructors
+
+  def this(child: Expression) = {
+    this(child, Literal(HllSketch.DEFAULT_LG_K), 0, 0)
+  }
+
+  def this(child: Expression, lgMaxK: Expression) = {
+    this(child, lgMaxK, 0, 0)
+  }
+
+  def this(child: Expression, lgMaxK: Int) = {
+    this(child, Literal(lgMaxK), 0, 0)
+  }
+
+  // Copy constructors required by ImperativeAggregate
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int):
+  HllUnionAgg = copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): HllUnionAgg =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  override protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression):
+  HllUnionAgg = copy(child = newLeft, lgMaxKExpression = newRight)
+
+  // Overrides for BinaryLike
+
+  override def left: Expression = child
+
+  override def right: Expression = lgMaxKExpression
+
+  // Overrides for TypedImperativeAggregate
+
+  override def prettyName: String = "hll_union_agg"
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(BinaryType, IntegerType)
+
+  override def dataType: DataType = BinaryType
+
+  override def nullable: Boolean = false
+
+  /**
+   * Instantiate an Union instance using the lgMaxK param.
+   *
+   * @return an Union instance
+   */
+  override def createAggregationBuffer(): Union = {
+    new Union(lgMaxK)
+  }
+
+  /**
+   * Update the Union instance with the HllSketch byte array obtained from the row.
+   *
+   * @param union The Union instance.
+   * @param input an input row
+   */
+  override def update(union: Union, input: InternalRow): Union = {
+    val v = child.eval(input)
+    if (v != null) {
+      child.dataType match {
+        case BinaryType =>
+          union.update(HllSketch.wrap(Memory.wrap(v.asInstanceOf[Array[Byte]])))
+        case _ => throw new UnsupportedOperationException(
+          s"A Union instance can only be updated with a valid HllSketch byte array")
+      }
+    }
+    union
+  }
+
+  /**
+   * Merges an input Union into the union which is acting as the aggregation buffer.
+   *
+   * @param union the Union instance used to store the aggregation result.
+   * @param input an input Union instance
+   */
+  override def merge(union: Union, input: Union): Union = {
+    union.update(input.getResult)
+    union
+  }
+
+  /**
+   * Returns an HllSketch derived from the merged HllSketches
+   *
+   * @param union Union instance used as an aggregation buffer
+   * @return A binary sketch which can be evaluated or merged
+   */
+  override def eval(union: Union): Any = {
+    union.toCompactByteArray
+  }
```

Comment:
Same question here: why compact representation?

Reply:
See my comment above, happy to change if you feel strongly.